plugin/kotel/tracer.go

```diff
@@ -0,0 +1,183 @@
+func TracerPropagator(propagator propagation.TextMapPropagator) TracerOpt {
+	return tracerOptFunc(func(t *Tracer) {
+		if t != nil {
+			t.propagators = propagator
+		}
+	})
+}
+
+// ClientID sets the optional client_id attribute value
+func ClientID(id string) TracerOpt {
+	return tracerOptFunc(func(t *Tracer) {
+		if t != nil {
+			t.clientID = id
+		}
+	})
+}
+
+// ConsumerGroup sets the optional group attribute value
+func ConsumerGroup(group string) TracerOpt {
+	return tracerOptFunc(func(t *Tracer) {
+		if t != nil {
+			t.consumerGroup = group
+		}
+	})
+}
+
+// Hooks ---------------------------------------------------------------------
+
+// OnProduceRecordBuffered starts a new span for the "send" operation on a buffered record
+//
+// It sets the attributes and injects the span context into the record's context
+func (t *Tracer) OnProduceRecordBuffered(r *kgo.Record) {
+	attrs := []attribute.KeyValue{
+		semconv.MessagingSystemKey.String("kafka"),
+		semconv.MessagingDestinationKindTopic,
+		semconv.MessagingDestinationKey.String(r.Topic),
+	}
+
+	if r.Key != nil {
+		attrs = append(attrs, semconv.MessagingKafkaMessageKeyKey.String(string(r.Key)))
+	}
+
+	if t.clientID != "" {
+		attrs = append(attrs, semconv.MessagingKafkaClientIDKey.String(t.clientID))
+	}
+
+	opts := []trace.SpanStartOption{
+		trace.WithAttributes(attrs...),
+		trace.WithSpanKind(trace.SpanKindProducer),
+	}
+
+	spanContext, _ := t.tracer.Start(r.Context, fmt.Sprintf("%s send", r.Topic), opts...)
+	t.propagators.Inject(spanContext, NewRecordCarrier(r))
+	r.Context = spanContext
+}
+
+// OnProduceRecordUnbuffered continues and ends the "send" span for an unbuffered record
+//
+// It sets attributes and records any error that occurred during the send operation
+func (t *Tracer) OnProduceRecordUnbuffered(r *kgo.Record, err error) {
+	span := trace.SpanFromContext(r.Context)
+	defer span.End()
+
+	span.SetAttributes(
+		semconv.MessagingKafkaPartitionKey.Int64(int64(r.Partition)),
+	)
+
+	if err != nil {
+		span.SetStatus(codes.Error, err.Error())
+		span.RecordError(err)
+	}
+}
+
+// OnFetchRecordBuffered starts and ends a new span for the "receive" operation on a buffered record
+//
+// It sets attributes and injects the span context into the record's context
+func (t *Tracer) OnFetchRecordBuffered(r *kgo.Record) {
+	attrs := []attribute.KeyValue{
+		semconv.MessagingSystemKey.String("kafka"),
+		semconv.MessagingDestinationKindTopic,
+		semconv.MessagingDestinationKey.String(r.Topic),
+		semconv.MessagingOperationReceive,
+		semconv.MessagingKafkaPartitionKey.Int64(int64(r.Partition)),
+	}
+
+	if r.Key != nil {
+		attrs = append(attrs, semconv.MessagingKafkaMessageKeyKey.String(string(r.Key)))
+	}
+
+	if t.clientID != "" {
+		attrs = append(attrs, semconv.MessagingKafkaClientIDKey.String(t.clientID))
+	}
+
+	if t.consumerGroup != "" {
+		attrs = append(attrs, semconv.MessagingKafkaConsumerGroupKey.String(t.consumerGroup))
+	}
+
```

Comment:
as a very minor nit, I'd delete the spaces around these conditionals -- it adds a lot of vertical space to this function (similar thought in other functions).

optional comment though, this is more of a style preference comment